src/rpc/fam_rpc_service_impl.cpp

```diff
@@ -718,33 +739,99 @@ int Fam_Rpc_Service_Impl::deregister_memory(uint64_t regionId,
     uint64_t dataitemId = offset / MIN_OBJ_SIZE;
     uint64_t rKey = generate_access_key(regionId, dataitemId, 0);
     uint64_t rwKey = generate_access_key(regionId, dataitemId, 1);
+    Fam_Region_Map_t *fiRegionMap;
 
-    pthread_mutex_lock(famOps->get_mr_lock());
-    auto rMr = fiMrs->find(rKey);
-    auto rwMr = fiMrs->find(rwKey);
-    if (rMr != fiMrs->end()) {
+    if (fiMrs == NULL)
+        fiMrs = famOps->get_fiMrs();
+
+    // Take read lock on fiMrs
+    pthread_rwlock_rdlock(famOps->get_mr_lock());
+    auto regionMrObj = fiMrs->find(regionId);
+    if (regionMrObj == fiMrs->end()) {
+        pthread_rwlock_unlock(famOps->get_mr_lock());
+        return 0;
+    } else {
+        fiRegionMap = regionMrObj->second;
+    }
+
+    // Take a writelock on fiRegionMap
+    pthread_rwlock_wrlock(&fiRegionMap->fiRegionLock);
+    // Release lock on fiMrs
+    pthread_rwlock_unlock(famOps->get_mr_lock());
+
+    auto rMr = fiRegionMap->fiRegionMrs->find(rKey);
+    auto rwMr = fiRegionMap->fiRegionMrs->find(rwKey);
+    if (rMr != fiRegionMap->fiRegionMrs->end()) {
         ret = fabric_deregister_mr(rMr->second);
         if (ret < 0) {
-            pthread_mutex_unlock(famOps->get_mr_lock());
+            pthread_rwlock_unlock(&fiRegionMap->fiRegionLock);
+            cout << "error: memory deregister failed" << endl;
```

Comment:
This is a case where memory server is in inconsistent state? may be better to stop memory server here or dont free the data item back to NVMM

Reply:
Talking about this, 
The current code path is :
1. fam_deallocate
2. memoryserver_allocator::deallocate will call heap free
3. calls unregister_memory

After step 2, NVMM can allocate same dataitem again. To avoid clash of register memory and unregister memory calls, it will be better if we move step 3 before step 2. I will raise an issue for this.

Reply:
I will not change this order now. May be we will address this issue later.